django/db/models/expressions.py

```diff
@@ -483,7 +485,14 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        return '%s', [self.value]
+        val = self.value
+        if self._output_field_or_none is not None:
+            from django.db.models.sql.compiler import SQLUpdateCompiler
+            if isinstance(compiler, SQLUpdateCompiler):
```

Comment:
A test case for DecimalField will help to narrow down what will be needed. The get_db_prep_save is going to be problematic for Value(decimal) though, since we don't require (and recently have actively recommended against) setting a max_digits/decimal_places. See https://github.com/django/django/commit/267a1dcd9b7877d97917ba6222fd247da060f9b0

If we have to differentiate between save and select, then we should try to pass that information via resolve_ref. Any test cases that will not pass due to decisions made earlier should be pulled out, and tickets created in trac with those test cases attached. These can be fixed after the alpha freeze, since they are not caused by your patch here.